File: library/src/main/java/com/github/library/utils/RequestParams.java
package com.github.library.utils;


import com.github.library.enums.RequestBodyType;
import com.github.library.helper.LogHelper;
import com.github.library.helper.general;
import com.github.library.model.FileModel;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Objects;

import okhttp3.FormBody;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;


/**
 * Created by alishatergholi on 2/21/18.
 */
public class RequestParams {

    private String plainText;

    private String contentType;

    private byte[] binaryInput;

    private LinkedHashMap<String, Object> params = new LinkedHashMap<>();

    private LinkedHashMap<String, FileModel> fileParams = new LinkedHashMap<>();

    private LinkedHashMap<String, List<Object>> paramsArray = new LinkedHashMap<>();

    private RequestBodyType type;

    private LogHelper logHelper = new LogHelper(RequestBody.class);

    public RequestParams() {
        this.type = RequestBodyType.FormUrlEncode;
    }

    public RequestParams(RequestBodyType type) {
        this.type = type;
    }

    public void put(String plainText) {
        this.plainText = plainText;
    }

    public void put(String contentType, byte[] binaryInput) {
        this.contentType = contentType;
        this.binaryInput = binaryInput;
    }

    public void put(String key, Object value) {
        if (general.StringIsEmptyOrNull(key)) {
            return;
        }
        params.put(key, value);
    }

    public void put(String key, String value) {
        if (general.StringIsEmptyOrNull(key) || general.StringIsEmptyOrNull(value)) {
            return;
        }
        params.put(key, value);
    }

    public void put(String key, Integer value) {
        if (general.StringIsEmptyOrNull(key) || value == null) {
            return;
        }
        params.put(key, value);
    }

    public void put(String key, Long value) {
        if (general.StringIsEmptyOrNull(key) || value == null) {
            return;
        }
        params.put(key, value);
    }

    public void put(String key, Float value) {
        if (general.StringIsEmptyOrNull(key) || value == null) {
            return;
        }
        params.put(key, value);
    }

    public void put(String key, Double value) {
        if (general.StringIsEmptyOrNull(key) || value == null) {
            return;
        }
        params.put(key, value);
    }

    public void put(String key, FileModel value) {
        if (general.StringIsEmptyOrNull(key) || value == null) {
            return;
        }
        fileParams.put(key, value);
    }

    public void put(String key, boolean value) {
        if (general.StringIsEmptyOrNull(key)) {
            return;
        }
        params.put(key, value);
    }

    public void put(String key, JSONObject value) {
        if (general.StringIsEmptyOrNull(key) || value == null) {
            return;
        }
        params.put(key, value);
    }

    public void put(String key, List<Object> values) {
        if (general.StringIsEmptyOrNull(key) || values == null) {
            return;
        }
        if (paramsArray.get(key) != null) {
            List<Object> items = paramsArray.get(key);
            if (items != null)
                values.addAll(items);
        }
        paramsArray.put(key, values);
    }

    public void add(String key, List<Object> values) {
        if (general.StringIsEmptyOrNull(key) || values == null) {
            return;
        }
        if (paramsArray.get(key) != null) {
            values.addAll(Objects.requireNonNull(paramsArray.get(key)));
        }
        paramsArray.put(key, values);
    }

    public void putContent(String key, List<Object> values) {
        if (general.StringIsEmptyOrNull(key) || values == null) {
            return;
        }
        ArrayList<Object> items = new ArrayList<>();
        if (paramsArray.get(key) != null) {
            items.addAll(Objects.requireNonNull(paramsArray.get(key)));
        }
        for (Object item : values) {
            items.add(item.toString());
        }
        paramsArray.put(key, items);
    }

    public void add(String key, String value) {
        if (general.StringIsEmptyOrNull(key) || value == null) {
            return;
        }
        ArrayList<Object> values;
        if (paramsArray.get(key) != null) {
            values = (ArrayList<Object>) paramsArray.get(key);
            if (values != null) {
                values.add(value);
            }
        } else {
            values = new ArrayList<>();
            values.add(value);
        }
        paramsArray.put(key, values);
    }

    public RequestBody getRequestForm() {
        if (getFileParams().size() > 0) {
            //for post File You should choose Multipart Type
            type = RequestBodyType.MultiPart;
        }
        switch (type) {
            case RawTEXTPlain:
                return RequestBody.create(MediaType.parse("text/plain; charset=utf-8"), plainText);
            case RawJSON:
                JSONObject object = new JSONObject();
                try {
                    for (String key : getParams().keySet()) {
                        object.put(key, getParams().get(key));
                    }
                    for (String key : getParamsArray().keySet()) {
                        for (Object item : Objects.requireNonNull(getParamsArray().get(key))) {
                            object.put(key, item);
                        }
                    }
                } catch (JSONException e) {
                    logHelper.e(e);
                }
                return RequestBody.create(MediaType.parse("application/json; charset=utf-8"), object.toString());
            case Binary:
                return RequestBody.create(MediaType.parse(contentType), binaryInput);
            case MultiPart:
                MultipartBody.Builder requestBody = new MultipartBody.Builder().setType(MultipartBody.FORM);
                for (String key : getParams().keySet()) {
                    requestBody.addFormDataPart(key, (String) Objects.requireNonNull(getParams().get(key)));
                }
                for (String key : getParamsArray().keySet()) {
                    for (Object item : Objects.requireNonNull(getParamsArray().get(key))) {
                        requestBody.addFormDataPart(key, (String) item);
                    }
                }
                for (String key : getFileParams().keySet()) {
                    FileModel uploadFile = fileParams.get(key);
                    if (uploadFile != null) {
                        requestBody.addFormDataPart(key, uploadFile.getFileName(), RequestBody.create(uploadFile.getMimeType(), uploadFile.getFile()));
                    }
                }
                return requestBody.build();
            case FormData:
            case FormUrlEncode:
            default: {
                FormBody.Builder formBody = new FormBody.Builder();
                for (String key : params.keySet()) {

                    formBody.addEncoded(key, String.valueOf(params.get(key)));
                }
                for (String key : paramsArray.keySet()) {
                    for (Object item : Objects.requireNonNull(paramsArray.get(key))) {
                        formBody.addEncoded(key, item.toString());
                    }
                }
                return formBody.build();
            }
        }
    }

    public RequestBodyType getType() {
        return type;
    }

    private LinkedHashMap<String, Object> getParams() {
        return params;
    }

    private LinkedHashMap<String, FileModel> getFileParams() {
        return fileParams;
    }

    private LinkedHashMap<String, List<Object>> getParamsArray() {
        return paramsArray;
    }

    public void setParamsArray(LinkedHashMap<String, List<Object>> paramsArray) {
        if (paramsArray == null) {
            return;
        }
        this.paramsArray = paramsArray;
    }
}

File: library/src/main/java/com/github/library/utils/RequestParams.kt
package com.github.library.utils


import com.github.library.enums.RequestBodyType
import com.github.library.helper.LogHelper
import com.github.library.helper.General
import com.github.library.model.FileModel

import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList

import java.util.Objects

import okhttp3.FormBody
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody


/**
 * Created by alishatergholi on 2/21/18.
 */
class RequestParams {

    private var plainText: String? = null

    private var contentType: String? = null

    private var binaryInput: ByteArray? = null

    private val params = LinkedHashMap<String, Any>()

    private val fileParams = HashMap<String, FileModel>()

    private var paramsArray: HashMap<String, List<Any>> = HashMap()

//    private var paramsArray: LinkedHashMap<String, List<Any>>? = LinkedHashMap()
//        set(paramsArray) {
//            if (paramsArray == null) {
//                return
//            }
//            field = paramsArray
//        }

    var type: RequestBodyType? = null
        private set

    private val logHelper = LogHelper(RequestBody::class.java)

    //for post File You should choose Multipart Type
    val requestForm: RequestBody
        get() {
            if (fileParams.size > 0) {
                type = RequestBodyType.MultiPart
            }
            when (type) {
                RequestBodyType.RawTEXTPlain -> return RequestBody.create(MediaType.parse("text/plain; charset=utf-8"), plainText!!)
                RequestBodyType.RawJSON -> {
                    val `object` = JSONObject()
                    try {
                        for (key in params.keys) {
                            `object`.put(key, params[key])
                        }
                        for (key in paramsArray.keys) {
                            for (item in Objects.requireNonNull<List<Any>>(paramsArray.get(key))) {
                                `object`.put(key, item)
                            }
                        }
                    } catch (e: JSONException) {
                        logHelper.e(e)
                    }

                    return RequestBody.create(MediaType.parse("application/json; charset=utf-8"), `object`.toString())
                }
                RequestBodyType.Binary -> return RequestBody.create(MediaType.parse(contentType!!), binaryInput!!)
                RequestBodyType.MultiPart -> {
                    val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
                    for (key in params.keys) {
                        builder.addFormDataPart(key, Objects.requireNonNull<Any>(params[key]) as String)
                    }
                    for (key in paramsArray.keys) {
                        for (item in Objects.requireNonNull<List<Any>>(paramsArray[key])) {
                            builder.addFormDataPart(key, item as String)
                        }
                    }
                    val fileParams = fileParams
                    for (key in fileParams.keys) {
                        val uploadFile = fileParams[key]
                        if (uploadFile != null) {
                            builder.addFormDataPart(key, uploadFile.fileName, RequestBody.create(uploadFile.mimeType, uploadFile.file!!))
                        }
                    }
                    return builder.build()
                }
                RequestBodyType.FormData, RequestBodyType.FormUrlEncode -> {
                    val formBody = FormBody.Builder()
                    for (key in params.keys) {

                        formBody.addEncoded(key, params[key].toString())
                    }
                    for (key in this.paramsArray.keys) {
                        for (item in Objects.requireNonNull<List<Any>>(this.paramsArray[key])) {
                            formBody.addEncoded(key, item.toString())
                        }
                    }
                    return formBody.build()
                }
                else -> {
                    val formBody = FormBody.Builder()
                    for (key in params.keys) {
                        formBody.addEncoded(key, params[key].toString())
                    }
                    for (key in this.paramsArray.keys) {
                        for (item in Objects.requireNonNull<List<Any>>(this.paramsArray[key])) {
                            formBody.addEncoded(key, item.toString())
                        }
                    }
                    return formBody.build()
                }
            }
        }

    constructor() {
        this.type = RequestBodyType.FormUrlEncode
    }

    constructor(type: RequestBodyType) {
        this.type = type
    }

    fun put(plainText: String) {
        this.plainText = plainText
    }

    fun put(contentType: String, binaryInput: ByteArray) {
        this.contentType = contentType
        this.binaryInput = binaryInput
    }

    fun put(key: String, value: Any) {
        if (General.stringIsEmptyOrNull(key)) {
            return
        }
        params[key] = value
    }

    fun put(key: String, value: String) {
        if (General.stringIsEmptyOrNull(key) || General.stringIsEmptyOrNull(value)) {
            return
        }
        params[key] = value
    }

    fun put(key: String, value: Int?) {
        if (General.stringIsEmptyOrNull(key) || value == null) {
            return
        }
        params[key] = value
    }

    fun put(key: String, value: Long?) {
        if (General.stringIsEmptyOrNull(key) || value == null) {
            return
        }
        params[key] = value
    }

    fun put(key: String, value: Float?) {
        if (General.stringIsEmptyOrNull(key) || value == null) {
            return
        }
        params[key] = value
    }

    fun put(key: String, value: Double?) {
        if (General.stringIsEmptyOrNull(key) || value == null) {
            return
        }
        params[key] = value
    }

    fun put(key: String, value: FileModel?) {
        if (General.stringIsEmptyOrNull(key) || value == null) {
            return
        }
        fileParams[key] = value
    }

    fun put(key: String, value: Boolean) {
        if (General.stringIsEmptyOrNull(key)) {
            return
        }
        params[key] = value
    }

    fun put(key: String, value: JSONObject?) {
        if (General.stringIsEmptyOrNull(key) || value == null) {
            return
        }
        params[key] = value
    }

    fun put(key: String, values: MutableList<Any>?) {
        if (General.stringIsEmptyOrNull(key) || values == null) {
            return
        }
        if (this.paramsArray[key] != null) {
            val items = this.paramsArray[key]
            if (items != null)
                values.addAll(items)
        }
        this.paramsArray[key] = values
    }

    fun add(key: String, values: MutableList<Any>?) {
        if (General.stringIsEmptyOrNull(key) || values == null) {
            return
        }
        if (this.paramsArray[key] != null) {
            values.addAll(this.paramsArray[key]!!)
        }
        this.paramsArray[key] = values
    }

    fun putContent(key: String, values: List<Any>?) {
        if (General.stringIsEmptyOrNull(key) || values == null) {
            return
        }
        val items = ArrayList<Any>()
        if (this.paramsArray[key] != null) {
            items.addAll(this.paramsArray[key]!!)
        }
        for (item in values) {
            items.add(item.toString())
        }
        this.paramsArray[key] = items
    }

    fun add(key: String, value: String?) {
        if (General.stringIsEmptyOrNull(key) || value == null) {
            return
        }
        var values: List<Any> = arrayListOf()
        if (this.paramsArray[key] != null) {
            values = this.paramsArray[key] as List<Any>
            value.plus(value)
        } else {
            values = ArrayList()
            values.add(value)
        }
        this.paramsArray[key] = values


    }
}

